hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/ContainerBalancer.java

```diff
@@ -839,20 +850,34 @@ private void resetState() {
   @Override
   public void notifyStatusChanged() {
     if (!scmContext.isLeader() || scmContext.isInSafeMode()) {
-      if (isBalancerRunning()) {
+      boolean shouldStop = false;
+
+      // lock here to ensure no change is made to the balancing thread while
+      // we're reading it
+      lock.lock();
+      try {
+        shouldStop = isBalancerRunning();
```

Comment:
I would suggest merging the two lock sections and afterwards stop can be called.